### ISSUE TITLE
Document /errorlog file format

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#3670]
The csc and vbc command lines now include an /errorlog option for specifying a file to which diagnostic information is written. The output format is some kind of json.

Please document the precise format of this file. I suggest adding the documentation to `docs/compilers` in markdown format.

**Comment by mavasani** [comment]
Yes, I'll take up this work item.

**Comment by mavasani** [comment]
We also need to document /reportanalyzer format. @gafter do you want to open a separate item to track that documentation or use this one itself?

**Comment by gafter** [comment]
@mavasani You're welcome to handle it either way is most convenient for you. If you're going to do them both at the same time, by all means use this bug. If not please open a separate issue.

**Comment by mavasani** [comment]
Adding @lgolding 

Larry, do you propose we document the currently emitted errorlog format for 1.1 OR should we wait for SARIF format to be finalized and implemented for compilers before documenting it?

If it is the latter, I will create a separate issue to track documenting /reportanalyzer switch, otherwise I will track both of them with this issue.

**Comment by lgolding** [comment]
@mavasani Since the compiler does emit this format, IMO we must document it.

**Comment by michaelcfanning** [comment]
I don't think we should document this format until it stabilizes. We should press to stabilize the format as soon as possible, in context of the SARIF standardization process, and then document.

**Comment by lgolding** [comment]
Michael explained this to me and I agree with him.

**Comment by mavasani** [comment]
This issue now tracks documenting just the /errorlog format, which will happen post 1.1

https://github.com/dotnet/roslyn/issues/6149 tracks documenting /reportanalyzer, which will happen before 1.1

**Comment by nguerrera** [comment]
@lgolding @michaelcfanning Should we not at least have a document linking to the SARIF spec in its pre-standardized form (https://github.com/sarif-standard/sarif-spec)? We could also use that to specify the structure of custom elements (e.g. "properties") used by Roslyn?

Also, I'm looking at what changes are needed to bring Roslyn up to the latest spec draft. Is there a known revision of the SARIF spec repo that corresponds to what was implemented in Roslyn? Would it make sense to retroactively declare that revision as 0.1 and anything beyond it as 0.2?

**Comment by lgolding** [comment]
It certainly makes sense to create a document that (1) points to the SARIF spec-in-progress, and (2) documents Roslyn's custom properties. You could certainly create that document.

The version of SARIF implemented in Roslyn was called "OMSTAR" (the "One Microsoft Static Analysis Results" format), which was the name we used before we decided to standardize it beyond Microsoft. There is a document that describes that format, and I can never remember where it is... hang on and I'll find it.

**Comment by lgolding** [comment]
... ah, [here it is](https://microsoft-my.sharepoint.com/personal/lgolding_microsoft_com/_layouts/15/WopiFrame.aspx?sourcedoc={2248375F-F5F2-4599-965D-FD075FB89713}&file=Static%20Analysis%20Results%20Interchange%20Format%20-%20external%20version.docx&action=default).

**Comment by lgolding** [comment]
... no wait a minute. @mavasani, weren't you going to take that document and put it in some public place?

**Comment by mavasani** [comment]
+ @srivatsn 

@lgolding I haven't done so, I thought as per [this](https://github.com/dotnet/roslyn/issues/3670#issuecomment-146958188) comment that we wanted to finalize the SARIF spec, move Roslyn compilers to that version and then document the format being SARIF. If we now feel that we should document current format, then I can put it on the docs section in this repo that is what we want.

By the way, Roslyn's error log output is a subset of the original OMSTAR spec - there were lot of things in there that we never emit. If we decide to document Roslyn compiler's current error log output format - we should consider documenting the format that it outputs, not the OMSTAR format.

**Comment by nguerrera** [comment]
Right now, I'm just trying to ascertain the delta between what's implemented and the current SARIF draft and figured older drafts of the spec might be informative vs. purely inspecting the implementation and empirical output.

**Comment by lgolding** [comment]
Oh right, I remember Michael saying that now. Nick, just curious what your current interest is in the format. I'm happy to share the doc with you.

**Comment by nguerrera** [comment]
I'm moving over to Michael's team and he asked me to look at bringing Roslyn up to current spec draft. :)

**Comment by lgolding** [comment]
Oh, that's great! In that case for sure you'll want to see the old document. You can schedule some time with me and I'll point out the signification changes. The documents look more different than the specs actually are, because I rewrote it in ISO spec-ese.

**Comment by mavasani** [comment]
Btw Nick, in case you didn't know, [here](http://source.roslyn.io/Microsoft.CodeAnalysis/R/f724fbca034fb970.html) is the implementation of the compiler's error logger and [here](http://source.roslyn.io/#q=ErrorLoggerTests) are some tests.
It's a pretty straightforward implementation actually, it might be easier to just read the code.

**Comment by nguerrera** [comment]
Thanks. I did find that already, just figured the docs couldn't hurt for additional context. :)

**Comment by nguerrera** [comment]
@mvasani, if you want to assign this to me, I can put up some docs referencing the current SARIF spec as part of the work to conform to it.

**Comment by dbolkensteyn** [comment]
@gafter This change isn't part of VS2015 Update 1 as it got merged 2 days after.

However, when using VS2015 Update 1 to compile Roslyn itself with `ErrorLog` set, I do get the new JSON format with `runLogs` in it.

Do you know how this is happening? Is there a way to get the old JSON format without `runLogs` and URIs?

**Comment by nguerrera** [comment]
@dbolkensteyn The log format is subject to change until version: "1.0" is emitted. This is because the specification for the underlying format is still a draft. There is no way to get the old format when building with more recent bits (as the Roslyn itself does). You will ultimately want to use the SARIF SDK (https://github.com/Microsoft/sarif-sdk) to read these files, but no matter what, you will have to be prepared for breaking changes across releases until the SARIF format is declared final.

Also, make note of more changes coming: https://github.com/dotnet/roslyn/pull/8166

/cc @srivatsn @lgolding @michaelcfanning

**Comment by dbolkensteyn** [comment]
Thanks @nguerrera - indeed I've realized that Roslyn is compiled itself with a previous Rolsyn milestone C# compiler, on which it depends through NuGet (it initially was surpising to me to see a different format only for Rolyn) - https://github.com/dotnet/roslyn/blob/98447e47aedfb415827ad47cad66582f05780aeb/build/ToolsetPackages/project.json#L6

I've created a ticket on the SonarQube C# Plugin to support this new version of SARIF: https://jira.sonarsource.com/browse/SONARCS-572

I'll wait for your PR to get merged before starting to work on that - thanks for the heads up

**Comment by michaelcfanning** [comment]
@dbolkensteyn, we'll be glad to review your SARIF format, if you'd like. we have a vscode extension that provides for validation against the externally published JSON schema. we also have an experimental VS-integrated SARIF viewer that will populate the error list from the file, might also be useful. feel free to ping me directly on this if you'd like. @lgolding, fyi.
